extensions/shared/upgrade-nudge/index.jsx

```diff
@@ -51,19 +48,27 @@ const UpgradeNudge = ( { planName, planPathSlug, postId, postType } ) => (
 	</Warning>
 );
 
-export default withSelect( ( select, { plan: planSlug } ) => {
-	const plan = select( 'wordpress-com/plans' ).getPlan( planSlug );
+export default compose( [
+	withSelect( ( select, { plan: planSlug } ) => {
+		const plan = select( 'wordpress-com/plans' ).getPlan( planSlug );
 
-	// WP.com plan objects have a dedicated `path_slug` field, Jetpack plan objects don't
-	// For Jetpack, we thus use the plan slug with the 'jetpack_' prefix removed.
-	const planPathSlug = startsWith( planSlug, 'jetpack_' )
-		? planSlug.substr( 'jetpack_'.length )
-		: get( plan, [ 'path_slug' ] );
+		// WP.com plan objects have a dedicated `path_slug` field, Jetpack plan objects don't
+		// For Jetpack, we thus use the plan slug with the 'jetpack_' prefix removed.
+		const planPathSlug = startsWith( planSlug, 'jetpack_' )
+			? planSlug.substr( 'jetpack_'.length )
+			: get( plan, [ 'path_slug' ] );
 
-	return {
-		planName: get( plan, [ 'product_name_short' ] ),
-		planPathSlug,
-		postId: select( 'core/editor' ).getCurrentPostId(),
-		postType: select( 'core/editor' ).getCurrentPostType(),
-	};
-} )( UpgradeNudge );
+		return {
+			planName: get( plan, [ 'product_name_short' ] ),
+			planPathSlug,
+			postId: select( 'core/editor' ).getCurrentPostId(),
+			postType: select( 'core/editor' ).getCurrentPostType(),
+		};
+	} ),
+	withDispatch( ( dispatch, { planPathSlug, postId, postType } ) => ( {
```

Comment:
This is interesting.

We're successfully getting `planPathSlug` here -- even though this is `ownProps`, as opposed to the props computed by `withSelect`!

I tried playing with this, e.g. using `{ select }` as the third argument here, and to use that to obtain `pathPlanSlug`, or to move the `window.location.href =` setting to the component itself, but all the _correct_ ways to do this failed at `autosave()`ing -- I'd always get the prompt asking me whether I wanted to discard my local changes to the post.

...whereas with this _incorrect_ way, things seem to work :thinking:

Reply:
> I tried [...] to move the `window.location.href =` setting to the component itself

```diff
diff --git a/extensions/shared/upgrade-nudge/index.jsx b/extensions/shared/upgrade-nudge/index.jsx
index b1a564470..5db9906d7 100644
--- a/extensions/shared/upgrade-nudge/index.jsx
+++ b/extensions/shared/upgrade-nudge/index.jsx
@@ -5,6 +5,7 @@ import { get, startsWith } from 'lodash';
 import { __, sprintf } from '@wordpress/i18n';
 import { addQueryArgs } from '@wordpress/url';
 import { Button } from '@wordpress/components';
+import { Component } from '@wordpress/element';
 import { compose } from '@wordpress/compose';
 import { withDispatch, withSelect } from '@wordpress/data';
 import { Warning } from '@wordpress/editor';
@@ -18,35 +19,49 @@ import './store';
 
 import './style.scss';
 
-const getUpgradeUrl = ( { planPathSlug, postId, postType } ) =>
-	addQueryArgs( `https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`, {
-		redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
-	} );
+class UpgradeNudge extends Component {
+	upgrade = async () => {
+		const { autosave, planPathSlug, postId, postType } = this.props;
 
-const UpgradeNudge = ( { autosaveAndRedirectToUpgrade, planName } ) => (
-	<Warning
-		actions={ [
-			<Button onClick={ autosaveAndRedirectToUpgrade } target="_top" isDefault>
-				{ __( 'Upgrade', 'jetpack' ) }
-			</Button>,
-		] }
-		className="jetpack-upgrade-nudge"
-	>
-		<div className="jetpack-upgrade-nudge__info">
-			<Gridicon className="jetpack-upgrade-nudge__icon" icon="star" size={ 18 } />
-			<div>
-				<span className="jetpack-upgrade-nudge__title">
-					{ sprintf( __( 'This block is available under the %(planName)s Plan.', 'jetpack' ), {
-						planName,
-					} ) }
-				</span>
-				<span className="jetpack-upgrade-nudge__message">
-					{ __( 'It will be hidden from site visitors until you upgrade.', 'jetpack' ) }
-				</span>
-			</div>
-		</div>
-	</Warning>
-);
+		await autosave();
+
+		window.location.href = addQueryArgs(
+			`https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`,
+			{
+				redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
+			}
+		);
+	};
+
+	render() {
+		const { planName } = this.props;
+
+		return (
+			<Warning
+				actions={ [
+					<Button onClick={ this.upgrade } target="_top" isDefault>
+						{ __( 'Upgrade', 'jetpack' ) }
+					</Button>,
+				] }
+				className="jetpack-upgrade-nudge"
+			>
+				<div className="jetpack-upgrade-nudge__info">
+					<Gridicon className="jetpack-upgrade-nudge__icon" icon="star" size={ 18 } />
+					<div>
+						<span className="jetpack-upgrade-nudge__title">
+							{ sprintf( __( 'This block is available under the %(planName)s Plan.', 'jetpack' ), {
+								planName,
+							} ) }
+						</span>
+						<span className="jetpack-upgrade-nudge__message">
+							{ __( 'It will be hidden from site visitors until you upgrade.', 'jetpack' ) }
+						</span>
+					</div>
+				</div>
+			</Warning>
+		);
+	}
+}
 
 export default compose( [
 	withSelect( ( select, { plan: planSlug } ) => {
@@ -65,10 +80,9 @@ export default compose( [
 			postType: select( 'core/editor' ).getCurrentPostType(),
 		};
 	} ),
-	withDispatch( ( dispatch, { planPathSlug, postId, postType } ) => ( {
-		autosaveAndRedirectToUpgrade: async () => {
+	withDispatch( dispatch => ( {
+		autosave: async () => {
 			await dispatch( 'core/editor' ).autosave();
-			window.location.href = getUpgradeUrl( { planPathSlug, postId, postType } );
 		},
 	} ) ),
 ] )( UpgradeNudge );
```

Reply:
> I tried playing with this, e.g. using `{ select }` as the third argument here, and to use that to obtain `pathPlanSlug`

(As documented at https://developer.wordpress.org/block-editor/packages/packages-data/#withDispatch)

```diff
diff --git a/extensions/shared/upgrade-nudge/index.jsx b/extensions/shared/upgrade-nudge/index.jsx
index b1a564470..4e0b92244 100644
--- a/extensions/shared/upgrade-nudge/index.jsx
+++ b/extensions/shared/upgrade-nudge/index.jsx
@@ -18,11 +18,6 @@ import './store';
 
 import './style.scss';
 
-const getUpgradeUrl = ( { planPathSlug, postId, postType } ) =>
-	addQueryArgs( `https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`, {
-		redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
-	} );
-
 const UpgradeNudge = ( { autosaveAndRedirectToUpgrade, planName } ) => (
 	<Warning
 		actions={ [
@@ -52,23 +47,32 @@ export default compose( [
 	withSelect( ( select, { plan: planSlug } ) => {
 		const plan = select( 'wordpress-com/plans' ).getPlan( planSlug );
 
-		// WP.com plan objects have a dedicated `path_slug` field, Jetpack plan objects don't
-		// For Jetpack, we thus use the plan slug with the 'jetpack_' prefix removed.
-		const planPathSlug = startsWith( planSlug, 'jetpack_' )
-			? planSlug.substr( 'jetpack_'.length )
-			: get( plan, [ 'path_slug' ] );
-
 		return {
 			planName: get( plan, [ 'product_name_short' ] ),
-			planPathSlug,
-			postId: select( 'core/editor' ).getCurrentPostId(),
-			postType: select( 'core/editor' ).getCurrentPostType(),
 		};
 	} ),
-	withDispatch( ( dispatch, { planPathSlug, postId, postType } ) => ( {
+
+	withDispatch( ( dispatch, { planSlug }, { select } ) => ( {
 		autosaveAndRedirectToUpgrade: async () => {
+			const plan = select( 'wordpress-com/plans' ).getPlan( planSlug );
+			const postId = select( 'core/editor' ).getCurrentPostId();
+			const postType = select( 'core/editor' ).getCurrentPostType();
+
+			// WP.com plan objects have a dedicated `path_slug` field, Jetpack plan objects don't
+			// For Jetpack, we thus use the plan slug with the 'jetpack_' prefix removed.
+			const planPathSlug = startsWith( planSlug, 'jetpack_' )
+				? planSlug.substr( 'jetpack_'.length )
+				: get( plan, [ 'path_slug' ] );
+
+			const upgradeUrl = addQueryArgs(
+				`https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`,
+				{
+					redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
+				}
+			);
+
 			await dispatch( 'core/editor' ).autosave();
-			window.location.href = getUpgradeUrl( { planPathSlug, postId, postType } );
+			window.location.href = upgradeUrl;
 		},
 	} ) ),
 ] )( UpgradeNudge );
```

Reply:
I gave the recommended implementation a spin and it seems to work well:

```diff
diff --git a/extensions/shared/upgrade-nudge/index.jsx b/extensions/shared/upgrade-nudge/index.jsx
index b1a564470..aff965dd4 100644
--- a/extensions/shared/upgrade-nudge/index.jsx
+++ b/extensions/shared/upgrade-nudge/index.jsx
@@ -18,57 +18,56 @@ import './store';
 
 import './style.scss';
 
-const getUpgradeUrl = ( { planPathSlug, postId, postType } ) =>
-	addQueryArgs( `https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`, {
-		redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
-	} );
-
-const UpgradeNudge = ( { autosaveAndRedirectToUpgrade, planName } ) => (
-	<Warning
-		actions={ [
-			<Button onClick={ autosaveAndRedirectToUpgrade } target="_top" isDefault>
-				{ __( 'Upgrade', 'jetpack' ) }
-			</Button>,
-		] }
-		className="jetpack-upgrade-nudge"
-	>
-		<div className="jetpack-upgrade-nudge__info">
-			<Gridicon className="jetpack-upgrade-nudge__icon" icon="star" size={ 18 } />
-			<div>
-				<span className="jetpack-upgrade-nudge__title">
-					{ sprintf( __( 'This block is available under the %(planName)s Plan.', 'jetpack' ), {
-						planName,
-					} ) }
-				</span>
-				<span className="jetpack-upgrade-nudge__message">
-					{ __( 'It will be hidden from site visitors until you upgrade.', 'jetpack' ) }
-				</span>
+const UpgradeNudge = ( { autosaveAndRedirectToUpgrade, wpcomPlan } ) => {
+	const planName = get( wpcomPlan, [ 'product_name_short' ] );
+	return (
+		<Warning
+			actions={ [
+				<Button onClick={ autosaveAndRedirectToUpgrade } target="_top" isDefault>
+					{ __( 'Upgrade', 'jetpack' ) }
+				</Button>,
+			] }
+			className="jetpack-upgrade-nudge"
+		>
+			<div className="jetpack-upgrade-nudge__info">
+				<Gridicon className="jetpack-upgrade-nudge__icon" icon="star" size={ 18 } />
+				<div>
+					<span className="jetpack-upgrade-nudge__title">
+						{ sprintf( __( 'This block is available under the %(planName)s Plan.', 'jetpack' ), {
+							planName,
+						} ) }
+					</span>
+					<span className="jetpack-upgrade-nudge__message">
+						{ __( 'It will be hidden from site visitors until you upgrade.', 'jetpack' ) }
+					</span>
+				</div>
 			</div>
-		</div>
-	</Warning>
-);
-
+		</Warning>
+	);
+};
 export default compose( [
-	withSelect( ( select, { plan: planSlug } ) => {
-		const plan = select( 'wordpress-com/plans' ).getPlan( planSlug );
-
+	withSelect( ( select, { plan: planSlug } ) => ( {
+		wpcomPlan: select( 'wordpress-com/plans' ).getPlan( planSlug ),
+	} ) ),
+	withDispatch( ( dispatch, { plan: planSlug, wpcomPlan }, { select } ) => {
 		// WP.com plan objects have a dedicated `path_slug` field, Jetpack plan objects don't
 		// For Jetpack, we thus use the plan slug with the 'jetpack_' prefix removed.
 		const planPathSlug = startsWith( planSlug, 'jetpack_' )
 			? planSlug.substr( 'jetpack_'.length )
-			: get( plan, [ 'path_slug' ] );
+			: get( wpcomPlan, [ 'path_slug' ] );
+		const postId = select( 'core/editor' ).getCurrentPostId();
+		const postType = select( 'core/editor' ).getCurrentPostType();
+
+		const upgradeUrl = ( { planPathSlug, postId, postType } ) =>
+			addQueryArgs( `https://wordpress.com/checkout/${ getSiteFragment() }/${ planPathSlug }`, {
+				redirect_to: `/${ postType }/${ getSiteFragment() }/${ postId }`,
+			} );
 
 		return {
-			planName: get( plan, [ 'product_name_short' ] ),
-			planPathSlug,
-			postId: select( 'core/editor' ).getCurrentPostId(),
-			postType: select( 'core/editor' ).getCurrentPostType(),
+			autosaveAndRedirectToUpgrade: async () => {
+				await dispatch( 'core/editor' ).autosave();
+				window.location.href = upgradeUrl( { planPathSlug, postId, postType } );
+			},
 		};
 	} ),
-	withDispatch( ( dispatch, { planPathSlug, postId, postType } ) => ( {
-		autosaveAndRedirectToUpgrade: async () => {
-			await dispatch( 'core/editor' ).autosave();
-			window.location.href = getUpgradeUrl( { planPathSlug, postId, postType } );
-		},
-	} ) ),
 ] )( UpgradeNudge );
```

Reply:
> > This is interesting.
> > We're successfully getting `planPathSlug` here -- even though this is `ownProps`, as opposed to the props computed by `withSelect`!
> 
> The way this composes, the props from `withSelect` _are_ `ownProps`, right?
> 
> ![Screen Shot 2019-07-10 at 18 15 56](https://user-images.githubusercontent.com/841763/60987053-1c080480-a341-11e9-9981-fbe63a0bdf74.png)

Ah right. I kinda read https://developer.wordpress.org/block-editor/packages/packages-data/#withDispatch as if it indicated that the third arg (i.e. the registry, or rather, `{ select }`) should be used to compute props from state/select (that's https://github.com/Automattic/jetpack/pull/13004#discussion_r301558597) -- possibly biased by how Redux works:

> In the majority of cases, it will be sufficient to use only two first params
passed to `mapDispatchToProps` as illustrated in the previous example.
However, there might be some very advanced use cases where using the
`registry` object might be used as a tool to optimize the performance of
your component. Using `select` function from the registry might be useful
when you need to fetch some dynamic data from the store at the time when the
event is fired, but at the same time, you never use it to render your
component. In such scenario, you can avoid using the withSelect higher
order component to compute such prop, which might lead to unnecessary
re-renders of your component caused by its frequent value change.

Reply:
> I gave the recommended implementation a spin and it seems to work well: [...]

Thanks for trying this out! It looks like the crucial point is that the plan object is coming from `withSelect`. It seems to me that moving the `window.location.href =` setting logic to the component (as done in https://github.com/Automattic/jetpack/pull/13004#discussion_r301556671) should be equally valid, yet that didn't work.

Reply:
Anyway, given your observation (https://github.com/Automattic/jetpack/pull/13004#discussion_r302162140), it's valid to rely on props computed by `withSelect` in `withDispatch`, so I guess we can proceed to either merge this PR as-is, or to apply your patch (https://github.com/Automattic/jetpack/pull/13004#discussion_r302177482) before doing so.

Reply:
> It seems to me that moving the `window.location.href =` setting logic to the component (as done in [#13004 (comment)](https://github.com/Automattic/jetpack/pull/13004#discussion_r301556671)) should be equally valid, yet that didn't work.

Changes in https://github.com/Automattic/jetpack/pull/13004#discussion_r302177482 seems to indicate that we have access to the promise inside `withDispatch` when we `await dispatch( 'core/editor' ).autosave();`. I seem to recall that the promise was not returned from the connected dispatch prop inside the component. I haven't verified, but that may explain the difference.

---

> Anyway, given your observation ([#13004 (comment)](https://github.com/Automattic/jetpack/pull/13004#discussion_r302162140)), it's valid to rely on props computed by `withSelect` in `withDispatch`, so I guess we can proceed to either merge this PR as-is, or to apply your patch ([#13004 (comment)](https://github.com/Automattic/jetpack/pull/13004#discussion_r302177482)) before doing so.

Agreed, this works in my testing. I'll approve and you can proceed as you see fit.